testcases/misc_tests/multi_instance.c

```diff
@@ -0,0 +1,410 @@
+/*
+ * COPYRIGHT (c) International Business Machines Corp. 2020
+ *
+ * This program is provided under the terms of the Common Public License,
+ * version 1.0 (CPL-1.0). Any use, reproduction or distribution for this
+ * software constitutes recipient's acceptance of CPL-1.0 terms which can be
+ * found in the file LICENSE file or at
+ * https://opensource.org/licenses/cpl1.0.php
+ */
+
+/* File: multi_instance.c
+ *
+ * Test driver.  In-depth regression test for PKCS #11
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <memory.h>
+#include <unistd.h>
+
+#include <dlfcn.h>
+#include <sys/timeb.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+
+#include "pkcs11types.h"
+#include "regress.h"
+
+
+int do_GetFunctionList(void);
+
+CK_RV do_GenerateTokenRSAKeyPair(CK_SLOT_ID slot_id, CK_SESSION_HANDLE sess,
+                                 CK_BYTE *label, CK_ULONG bits,
+                                 CK_OBJECT_HANDLE *hPubKey,
+                                 CK_OBJECT_HANDLE *hPrivKey)
+{
+    CK_MECHANISM mech;
+    CK_RV rv;
+    CK_MECHANISM_INFO rsakeygeninfo;
+    CK_BBOOL false = 0;
+    CK_BYTE pub_exp[] = { 0x1, 0x0, 0x1 };
+    CK_ATTRIBUTE pub_tmpl[] = {
+        {CKA_MODULUS_BITS, &bits, sizeof(bits)},
+        {CKA_PUBLIC_EXPONENT, &pub_exp, sizeof(pub_exp)},
+        {CKA_LABEL, label, (CK_ULONG) strlen((char *) label) + 1},
+        {CKA_TOKEN, &false, sizeof(CK_BBOOL)}
+    };
+    CK_ATTRIBUTE priv_tmpl[] = {
+        {CKA_LABEL, label, (CK_ULONG) strlen((char *) label) + 1},
+        {CKA_TOKEN, &false, sizeof(CK_BBOOL)}
+    };
+
+    rv = funcs->C_GetMechanismInfo(slot_id, CKM_RSA_PKCS_KEY_PAIR_GEN,
+                                   &rsakeygeninfo);
+    if (rv != CKR_OK) {
+        testcase_fail("C_GetMechanismInfo(CKM_RSA_PKCS_KEY_PAIR_GEN) rc = %s", p11_get_ckr(rv));
+        return rv;
+    }
+
+    mech.mechanism = CKM_RSA_PKCS_KEY_PAIR_GEN;
+    mech.ulParameterLen = 0;
+    mech.pParameter = NULL;
+
+    rv = funcs->C_GenerateKeyPair(sess, &mech, pub_tmpl, 4, priv_tmpl, 2,
+                                  hPubKey, hPrivKey);
+    if (rv != CKR_OK) {
+        testcase_fail("C_GenerateKeyPair rc = %s", p11_get_ckr(rv));
+        return rv;
+    }
+    return CKR_OK;
+}
+
+
+int main(int argc, char **argv)
```

Comment:
Great test case as well. Could you add a RWEADME(.md) file in the test case subdir, which lists all tests which have to invoked by hand i.e., do not run with the test suite ? Im afraid w/o one, those special test cases are forgotten or run seldomly.

Reply:
Good idea. Will do.